### PR TITLE
Extend and add tests for factory models

### PR DESCRIPTION
**Pull request by eduardoj**
Co-authored-by: David Kang <dkang@suse.com>

### CODE REVIEW

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 1]
```diff
@@ -1,17 +1,229 @@
...
     it { expect(strategy_for('openSUSE:42.3')).to            be_kind_of ObsFactory::DistributionStrategyOpenSUSE }
     it { expect(strategy_for('SUSE:SLE-12-SP1:GA')).to       be_kind_of ObsFactory::DistributionStrategySLE12SP1 }
     it { expect(strategy_for('SUSE:SLE-15:GA')).to           be_kind_of ObsFactory::DistributionStrategySLE15 }
     it { expect(strategy_for('SUSE:SLE-15-SP1:GA')).to       be_kind_of ObsFactory::DistributionStrategySLE15 }
     it { expect(strategy_for('SUSE:SLE-12-SP3:Update:Products:CASP20')).to be_kind_of ObsFactory::DistributionStrategyCasp }
   end
+
+  describe 'self.attributes' do
+    it { expect(ObsFactory::Distribution.attributes.class).to eq(Array) }
+    it { expect(ObsFactory::Distribution.attributes.count).to eq(13) }
```
Comment (coolo) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: What is this testing? I find the 13 a bit magic here

File: src/api/.rubocop_todo.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # This configuration was generated by
 # `rubocop --auto-gen-config`
-# on 2018-06-27 11:19:32 +0000 using RuboCop version 0.57.2.
+# on 2018-06-28 15:12:27 +0000 using RuboCop version 0.57.2.
```
Comment (Ana06) [inline src/api/.rubocop_todo.yml]: why is that needed to write tests? :confused:

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 3]
```diff
@@ -1,17 +1,229 @@
...
     it { expect(strategy_for('openSUSE:Factory:PowerPC')).to be_kind_of ObsFactory::DistributionStrategyFactoryPPC }
     it { expect(strategy_for('openSUSE:42.3')).to            be_kind_of ObsFactory::DistributionStrategyOpenSUSE }
     it { expect(strategy_for('SUSE:SLE-12-SP1:GA')).to       be_kind_of ObsFactory::DistributionStrategySLE12SP1 }
     it { expect(strategy_for('SUSE:SLE-15:GA')).to           be_kind_of ObsFactory::DistributionStrategySLE15 }
     it { expect(strategy_for('SUSE:SLE-15-SP1:GA')).to       be_kind_of ObsFactory::DistributionStrategySLE15 }
     it { expect(strategy_for('SUSE:SLE-12-SP3:Update:Products:CASP20')).to be_kind_of ObsFactory::DistributionStrategyCasp }
   end
+
+  describe 'self.attributes' do
+    it { expect(ObsFactory::Distribution.attributes.class).to eq(Array) }
```
Comment (Ana06) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: why is it important that it returns an array?

File: src/api/spec/models/obs_factory/distribution_spec.rb (hunk 1 above)
Comment (Ana06) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: I agree with @coolo, why 13? I also don't understand what this test is really testing... :confused:

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 4]
```diff
@@ -1,17 +1,229 @@
...
+
+  describe '#staging_projects' do
+    context 'with staging projects' do
+      let!(:staging_project) { create(:project, name: 'openSUSE:Factory:Staging') }
+      let!(:staging_project_a) { create(:project, name: 'openSUSE:Factory:Staging:A') }
+      let!(:staging_project_b) { create(:project, name: 'openSUSE:Factory:Staging:B') }
+      let!(:staging_project_c) { create(:project, name: 'openSUSE:Factory:Staging:C') }
+      let!(:staging_project_345) { create(:project, name: 'openSUSE:Factory:Staging:345') }
+
+      it { expect(distribution.staging_projects.count).to eq(3) }
```
Comment (Ana06) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: why 3? or better, which 3?

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 5]
```diff
@@ -1,17 +1,229 @@
...
+
+  describe '#staging_projects_all' do
+    context 'with staging projects' do
+      let!(:staging_project) { create(:project, name: 'openSUSE:Factory:Staging') }
+      let!(:staging_project_a) { create(:project, name: 'openSUSE:Factory:Staging:A') }
+      let!(:staging_project_b) { create(:project, name: 'openSUSE:Factory:Staging:B') }
+      let!(:staging_project_c) { create(:project, name: 'openSUSE:Factory:Staging:C') }
+      let!(:staging_project_345) { create(:project, name: 'openSUSE:Factory:Staging:345') }
+
+      it { expect(distribution.staging_projects_all.count).to eq(4) }
```
Comment (Ana06) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: same here, shouldn't we test which 4? :confused:

File: src/api/spec/models/obs_factory/staging_project_spec.rb [hunk 6]
```diff
@@ -1,10 +1,24 @@
 require 'rails_helper'
+require 'webmock/rspec'
 
 RSpec.describe ObsFactory::StagingProject do
-  let(:distribution) { ObsFactory::Distribution.new(create(:project, name: 'openSUSE:Factory')) }
+  let(:factory) { create(:project, name: 'openSUSE:Factory') }
```
Comment (Ana06) [inline src/api/spec/models/obs_factory/staging_project_spec.rb]: why all this need to be always created? :confused:

File: src/api/spec/models/obs_factory/staging_project_spec.rb [hunk 7]
```diff
@@ -39,4 +53,623 @@
...
+
+      it { expect(subject.count).to eq(3) }
+    end
+
+    context 'openSUSE:13.2' do
+      let(:opensuse_132_distribution) { ObsFactory::Distribution.find(opensuse_132.name) }
+      subject { ObsFactory::StagingProject.for(opensuse_132_distribution) }
+
+      it { expect(subject.count).to eq(2) }
+    end
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/staging_project_spec.rb]: I find it quite hard to understand what #for does by reading through this test. Does the behaviour of this method change when using  13.2 instead of factory ppc?
If not I would use the same distro and make it more visible what causes the count / result list to change. If I understand correctly it's the subprojects

File: src/api/spec/models/obs_factory/staging_project_spec.rb [hunk 8]
```diff
@@ -39,4 +53,623 @@
...
+  end
+
+  describe '.name' do
+    subject { staging_project_a }
+
+    it { expect(subject.name).to eq('openSUSE:Factory:Staging:A') }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/staging_project_spec.rb]: I had to scroll up and down a couple of times to understand this test.

Please change it to something like this:

```ruby
subject { create(:obs_factory_staging_project, name: 'openSUSE:Factory:Staging:A') }

it { expect(subject.name).to eq('openSUSE:Factory:Staging:A') }
```

Comment (bgeuken) [reply src/api/spec/models/obs_factory/staging_project_spec.rb]: This would also come handy for many of the other tests below

**Comment by eduardoj** [comment]
@Ana06 

> maybe it would have been a good idea to split this PR.. the commits are completely independent...

I will divide this pull request, in two pull requests, at least.

File: src/api/spec/models/obs_factory/distribution_spec.rb (hunk 3 above)
Comment (DavidKang) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: we will remove it

**Comment by eduardoj** [comment]
@Ana06, @bgeuken, @coolo, model tests for staging_project where moved to #5261. Please have a look at them there.

**Comment by eduardoj** [comment]
@Ana06, @bgeuken, @coolo, required changes applied. File `.rubocop_todo.yml` moved to #5261. Please re-review again.

File: src/api/spec/models/obs_factory/distribution_spec.rb (hunk 3 above)
Comment (bgeuken) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: This is still there:-)

File: src/api/spec/models/obs_factory/distribution_spec.rb (hunk 1 above)
Comment (bgeuken) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: Same here

Comment (bgeuken) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: Actually, I don't think we need a test for this. Since [attributes](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/obs_factory/distribution.rb#L24) is just a static array.

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 9]
```diff
@@ -1,17 +1,229 @@
...
     it { expect(strategy_for('SUSE:SLE-12-SP3:Update:Products:CASP20')).to be_kind_of ObsFactory::DistributionStrategyCasp }
   end
+
+  describe 'self.attributes' do
+    it { expect(ObsFactory::Distribution.attributes.class).to eq(Array) }
+    it { expect(ObsFactory::Distribution.attributes.count).to eq(13) }
+  end
+
+  describe 'self.find' do
+    context 'with a project' do
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: I would define the project within that block. That way the test becomes much more readable.

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 10]
```diff
@@ -1,17 +1,229 @@
...
+      context 'with distribution' do
+        subject { ObsFactory::Distribution.find(project.name) }
+
+        it { expect(subject.class).to eq(ObsFactory::Distribution) }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: You could use the [be_kind_of](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/type-matchers) matcher.

Comment (bgeuken) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: I would add a second test here that ensures that the project was set.

File: src/api/spec/models/obs_factory/distribution_spec.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/spec/models/obs_factory/distribution_spec.rb]: I would add a label to this it block that explains what is expected.

And instead of counting you could match the arrays.

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 11]
```diff
@@ -1,17 +1,229 @@
...
+      let!(:staging_project_a) { create(:project, name: 'openSUSE:Factory:Staging:A') }
+      let!(:staging_project_b) { create(:project, name: 'openSUSE:Factory:Staging:B') }
+      let!(:staging_project_c) { create(:project, name: 'openSUSE:Factory:Staging:C') }
+      let!(:staging_project_345) { create(:project, name: 'openSUSE:Factory:Staging:345') }
+
+      it { expect(distribution.staging_projects.count).to eq(3) }
+    end
+
+    context 'without staging projects' do
+      it { expect(distribution.staging_projects.count).to eq(0) }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: it { expect(distribution.staging_projects).to be_empty }

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 12]
```diff
@@ -1,17 +1,229 @@
...
+      let!(:ring_project_1) { create(:project, name: 'openSUSE:Factory:Rings:1-MinimalX') }
+      let!(:ring_project_2) { create(:project, name: 'openSUSE:Factory:Rings:2-TestDVD') }
+      let(:ring_project_names) do
+        ['openSUSE:Factory:Rings:0-Bootstrap',
+         'openSUSE:Factory:Rings:1-MinimalX',
+         'openSUSE:Factory:Rings:2-TestDVD']
+      end
+
+      it { expect(distribution.ring_projects.count).to eq(3) }
+      it { expect(distribution.ring_projects.map(&:name)).to eq(ring_project_names) }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: Have you tried using [contain_exactly](https://relishapp.com/rspec/rspec-expectations/v/3-7/docs/built-in-matchers/contain-exactly-matcher)?

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 13]
```diff
@@ -21,8 +21,13 @@ def strategy_for(name)
...
   describe 'self.attributes' do
-    it { expect(ObsFactory::Distribution.attributes.class).to eq(Array) }
-    it { expect(ObsFactory::Distribution.attributes.count).to eq(13) }
+    let(:result) do
+      %w[name description staging_projects openqa_version openqa_group
+         source_version totest_version published_version staging_manager
+         standard_project live_project images_project ring_projects]
+    end
+
+    it { expect(ObsFactory::Distribution.attributes).to eq(result) }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: Please merge this into the first commit.

File: src/api/spec/models/obs_factory/distribution_spec.rb [hunk 14]
```diff
@@ -99,7 +108,11 @@ def strategy_for(name)
       let!(:staging_project_c) { create(:project, name: 'openSUSE:Factory:Staging:C') }
       let!(:staging_project_345) { create(:project, name: 'openSUSE:Factory:Staging:345') }
 
-      it { expect(distribution.staging_projects_all.count).to eq(4) }
+      let(:result) { [staging_project_a, staging_project_b, staging_project_c, staging_project_345] }
+
+      subject { distribution.staging_projects_all.map(&:project) }
+
+      it { expect(subject).to eq(result) }
```
Comment (bgeuken) [inline src/api/spec/models/obs_factory/distribution_spec.rb]: I would use match_array  or contain_exactly here.
